ISPN-10496 Support surefire rerunFailingTestsCount

diegolovison opened the pull request:
https://issues.jboss.org/browse/ISPN-10496

This PR address the fix for:
org.apache.maven.plugin.surefire.StartupReportConfiguration
```java
   public StatelessXmlReporter instantiateStatelessXmlReporter(Integer forkNumber) {
      assert forkNumber == null || this.isForkMode;

      Map<String, Map<String, List<WrappedReportEntry>>> testClassMethodRunHistory = this.isForkMode ? new ConcurrentHashMap() : this.testClassMethodRunHistory;
      return this.isDisableXmlReport() ? null : new StatelessXmlReporter(this.resolveReportsDirectory(forkNumber), this.reportNameSuffix, this.trimStackTrace, this.rerunFailingTestsCount, (Map)testClassMethodRunHistory, this.xsdSchemaLocation);
   }
```
The root cause is because pom.xml by default `<disableXmlReport>true</disableXmlReport>`

danberindei commented on server/runtime/pom.xml:
```diff
@@ -336,6 +336,7 @@
                <threadCount>${infinispan.test.parallel.threads}</threadCount>
                <forkCount>1</forkCount>
                <reuseForks>true</reuseForks>
+
```
I think you wanted to commit something else Diego :)

diegolovison replied:
ATM, no. My idea is to have https://ci.infinispan.org/job/Infinispan/job/PR-7215/1/testReport/ 
After that, I will submit my changes and see the results and start fixing. 
The big problem is that `<disableXmlReport>true</disableXmlReport>` must be false. And then without the initial report, I won't be able to know if I am doing some progress

danberindei replied:
I thought you'd just compare with master... Jenkins merges your branch with master before building anyway, so the number of tests in the latest build won't necessarily match the initial PR build.

Anyway, time to fix the duplicate test results now :)

Also have a look at the

danberindei replied:
So you want to keep `<disableXmlReport>true</disableXmlReport>` on the next line here?

diegolovison commented:
Even when `<testNGListeners></testNGListeners>` is empty `PolarionJUnitXMLReporter.java ` is running. I removed it.

diegolovison commented:
@danberindei it is ready for review.
Keep in mind that `-Dsurefire.rerunFailingTestsCount=5` is not working even when the Jenkins has https://plugins.jenkins.io/flaky-test-handler installed.

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/PolarionTestNameListenerAdapter.java:
```diff
@@ -0,0 +1,127 @@
+
+   @Override
+   public void onTestFailure(ITestResult result) {
+      super.onTestFailure(result);
```
No need to override all the methods if you only need `onTestStart()`

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/PolarionTestNameListenerAdapter.java:
```diff
@@ -0,0 +1,127 @@
+
+   private void fixTestName(ITestResult result) {
+      try {
+         Field xField = result.getClass().getDeclaredField("m_name");
```
It's a hack, but I guess it will work... you just need to obtain the field and cache it in a static block.

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/PolarionTestNameListenerAdapter.java:
```diff
@@ -0,0 +1,127 @@
+ * @author <a href='mailto:afield[at]redhat[dot]com'>Alan Field</a>
+ * @author <a href='mailto:dlovison[at]redhat[dot]com'>Diego Lovison</a>
+ */
+public class PolarionTestNameListenerAdapter extends TestListenerAdapter {
```
Do you really need to extend `TestListenerAdapter`? I don't think you need to keep track of all the tests that have run any more.

And longer javadoc to explain what you're doing and why might help in the future when we upgrade TestNG and this stops working.

danberindei replied:
@diegolovison did you forget to push the change to implement `IResultListener2` instead of extending `TestListenerAdapter`? You could also implement only `ITestListener`, it has fewer methods and I don't think you need anything from `IResultListener2`. Also the class name shouldn't end in `Adapter`, because it's just a listener, it's not adapting another listener.

diegolovison replied:
I implemented `ITestListener`. `Adapter` from the class name was removed.

diegolovison commented:
Hey, @danberindei I believe that I addressed all your comments.
My build has 26161 and [master_1438](https://ci.infinispan.org/job/Infinispan/job/master/1438) 26129 = 32

> org.infinispan.jcache contains 2 failures
> - org.infinispan.jcache.JCacheConfigurationTest.ThreadLeakChecker
> - org.infinispan.jcache.JCacheConfigurationTest.ThreadLeakChecker

= 30

> UNKNOWN[null].ThreadLeakChecker

= 29

> org.infinispan.client.hotrod

= 20

> org.infinispan.distribution.IllegalMonitorTest

= 0

**Additional Info:**

1)
**org.infinispan.client.hotrod** is duplicating the following entries
testIteration[DIST_SYNC]([ENTRYSET, 3.0])
testIteration[DIST_SYNC]([KEYSET, 3.0])
testIteration[DIST_SYNC]([VALUES, 3.0])
testIteration[LOCAL]([ENTRYSET, 3.0])
testIteration[LOCAL]([KEYSET, 3.0])
testIteration[LOCAL]([VALUES, 3.0])
testIteration[REPL_SYNC]([ENTRYSET, 3.0])
testIteration[REPL_SYNC]([KEYSET, 3.0])
testIteration[REPL_SYNC]([VALUES, 3.0])

2)
**org.infinispan.distribution.IllegalMonitorTest** is reporting one test for each iteration in the test.

-------

For item 1) and 2) I would like to address and fix in a diff PR. WDYT?

danberindei commented on Jenkinsfile:
```diff
@@ -56,7 +56,7 @@ pipeline {
             steps {
                 configFileProvider([configFile(fileId: 'maven-settings-with-deploy-snapshot', variable: 'MAVEN_SETTINGS')]) {
-                    sh "$MAVEN_HOME/bin/mvn verify -B -V -e -s $MAVEN_SETTINGS -Dmaven.test.failure.ignore=true -Dansi.strip=true"
+                    sh "$MAVEN_HOME/bin/mvn verify -B -V -e -s $MAVEN_SETTINGS -Dmaven.test.failure.ignore=true -Dansi.strip=true -Dsurefire.rerunFailingTestsCount=5"
```
Do you want this in master and PR builds? I thought you only wanted to be able to do it in your own builds.

diegolovison replied:
Yes. Because it will start reporting what are the flaky tests.

danberindei replied:
How will that happen? Is there any documentation that you could link to in a comment here?

diegolovison replied:
I would like to use https://plugins.jenkins.io/flaky-test-handler but it is not working. We need to spend more time but the first goal of this PR is to remove `<disableXmlReport>false</disableXmlReport>` that is the first restriction from the surefire maven plugin

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/CustomJUnitXMLWriter.java:
```diff
@@ -9,15 +9,15 @@
 
 /**
- * A JUnit XML report generator for Polarion based on the JUnitXMLReporter
+ * A JUnit XML report to prevent crashing the surefire fork JVM based on JUnitXMLReporter.
```
```suggestion
 * A JUnit XML report generator to report failures where throwing an exception would crash the surefire fork JVM.
```

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/CustomJUnitXMLWriter.java:
```diff
@@ -9,15 +9,15 @@
+ * A JUnit XML report to prevent crashing the surefire fork JVM based on JUnitXMLReporter.
  *
- * <p>Extracted from {@link PolarionJUnitXMLReporter}</p>
+ * This class is responsible to log hanged tests and leaked threads.
```
```suggestion
 * Based on JUnitXMLReporter.
```
Not really, it's responsible for reporting in the JUnit XML report format whatever other classes decide to report. For now, it's just leaked threads.

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/PolarionTestNameListener.java:
```diff
@@ -0,0 +1,70 @@
+/**
+ *
+ * This listener fixes the method name by overwriting the private `TestResult.m_name` field and performs some
+ * additional processing to ensure the resulting test name is unique.
```
The javadoc you posted in the chat had 2 more paragraphs

diegolovison replied:
The class was renamed to `TestNGTestNameListener`.  The javadoc was moved to `TestNGNameTranslator` that where is the hack is present.

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/PolarionTestNameListener.java:
```diff
@@ -0,0 +1,70 @@
+ * @author <a href='mailto:afield[at]redhat[dot]com'>Alan Field</a>
+ * @author <a href='mailto:dlovison[at]redhat[dot]com'>Diego Lovison</a>
+ */
+public class PolarionTestNameListener extends TestNGNameTranslator implements ITestListener {
```
Shouldn't you remove `Polarion` from the name here as well, if you did it for `PolarionJUnitXMLWriter`?

Also seems weird to extend `TestNGNameTranslator` just to invoke a method, when you could have made the method static and invoked it directly.

diegolovison replied:
The class was renamed to TestNGTestNameListener. Created a static method

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/TestNGNameTranslator.java:
```diff
@@ -0,0 +1,80 @@
+ */
+public class TestNGNameTranslator {
+
+   public String translanteTestName(ITestResult res) {
```
```suggestion
   public static String translateTestName(ITestResult res) {
```

diegolovison replied:
Done

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/TestNGNameTranslator.java:
```diff
@@ -0,0 +1,80 @@
+import org.testng.annotations.Test;
+
+/**
+ * When a test is annotated with {@code @Test(testName = "foo")}, TestNG reports all the test methods in the class as
```
```suggestion
 * When a test is annotated with {@code @Test(testName = "foo")}, {@code ITestResult.getTestName()} reports all the test methods in the class as
```

"reports" doesn't work very well here because it's no longer just fixing the XML reports.

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/TestNGNameTranslator.java:
```diff
@@ -0,0 +1,80 @@
+                  .append(" times");
+         }
+         // JCache tests are a special case
+         if (getModuleSuffix().contains("jcache")) {
```
I haven't noticed before, but this step should be moved outside the parent if (`if (res.getMethod().getConstructorOrMethod().getMethod().isAnnotationPresent(Test.class))`) because it doesn't work on the `@Test` annotation.

diegolovison replied:
Good catch but now we have 3 new classes `JUnitNameTranslator`, `JUnitTestNameListener`, `ModuleNameTranslator` and 1 new property `inner.infinispan.module-suffix` because JCache TCK tests are JUnit based.

Now the console will print the right module under the test but the XML still has the name of the method. `JUnitTestNameListener` is not working because Surefire is responsible to generate the XML and is using the method name.

diegolovison replied:
`JUnitTestNameListener` was removed and `reportNameSuffix` was introduced. It will add the suffix at the end of the class name:

```xml
  <testcase name="testEqualsForAccessedExpiryPolicy" classname="javax.cache.expiry.ExpiryPolicyTest(embedded)" time="0.001"/>
  <testcase name="testModifiedExpiryPolicy" classname="javax.cache.expiry.ExpiryPolicyTest(embedded)" time="0.021"/>
  <testcase name="testEqualsForModifiedExpiryPolicy" classname="javax.cache.expiry.ExpiryPolicyTest(embedded)" time="0.018"/>
```

I believe that once we have https://issues.apache.org/jira/browse/SUREFIRE-1696 fixed we could remove the attribute `JUnitTestListener#progressLogger`

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/TestNGTestListener.java:
```diff
@@ -1,21 +1,17 @@
  */
-public class TestNGTestListener implements ITestListener, IConfigurationListener2, ISuiteListener {
-   private static final Logger log = Logger.getLogger(TestNGTestListener.class);
+public class TestNGTestListener extends TestNGNameTranslator implements ITestListener, IConfigurationListener2,
```
-1 to extend

diegolovison commented:
@danberindei it is ready for one more round! Thanks for all your feedback.

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/JUnitTestListener.java:
```diff
@@ -60,8 +61,10 @@ public void testAssumptionFailure(Failure failure) {
       String className = description.isSuite() ? "suite" : description.getTestClass().getSimpleName();
-      return className + "." + description.getMethodName();
+      sb.append(className).append(".").append(translateTestName(description));
+      return sb.toString();
```
In JDK 11 `+` uses `invokedynamic` and at least in theory can be optimized better than an explicit `StringBuilder`.

danberindei commented on commons-test/src/main/java/org/infinispan/commons/test/TestNGTestListener.java:
```diff
@@ -64,17 +61,7 @@ public void onFinish(ITestContext context) {
-         }
-      }
-      return result.toString();
+      return new StringBuilder(res.getInstanceName()).append(".").append(translateTestName(res)).toString();
```
Regular `+` looks better and is at least as fast.

diegolovison commented:
Closing.. Let's try to remove testNG in ISPN11
cc @danberindei @tristantarrant

danberindei commented:
I'm all for it @diegolovison, but it's not going to be easy!

diegolovison commented:
@alanfx this could be a nice "big task" for QE

alanfx commented:
@diegolovison I'll add it to our TODO list!
